src/mlpack/methods/kde/kde_main.cpp

```diff
@@ -0,0 +1,165 @@
+PARAM_MATRIX_IN_REQ("query", "Query dataset to KDE on.", "q");
+PARAM_DOUBLE_IN("bandwidth", "Bandwidth of the kernel", "b", 1.0);
+
+// Load or save models.
+PARAM_MODEL_IN(KDEModel,
+               "input_model",
+               "File containing pre-trained KDE model.",
+               "m");
+PARAM_MODEL_OUT(KDEModel,
+                "output_model",
+                "If specified, the KDE model will be saved to the given file.",
+                "M");
+
+// Configuration options
+PARAM_STRING_IN("kernel", "Kernel to use for the estimation"
+    "('gaussian', 'epanechnikov').", "k", "gaussian");
+PARAM_STRING_IN("tree", "Tree to use for the estimation"
+    "('kd-tree', 'ball-tree').", "t", "kd-tree");
+PARAM_DOUBLE_IN("rel_error",
+                "Relative error tolerance for the result",
+                "e",
+                1e-8);
+PARAM_DOUBLE_IN("abs_error",
+                "Relative error tolerance for the result",
+                "E",
+                0.0);
+PARAM_FLAG("breadth_first", "Use breadth-first traversal instead of depth"
+           "first.", "w");
+// Maybe in the future it could be interesting to implement different metrics.
+
+// Output options.
+PARAM_MATRIX_OUT("output", "Matrix to store output estimations.",
+    "o");
+
+static void mlpackMain()
+{
+  const size_t output_precision = 40;
+  // Get some parameters.
+  arma::mat query = std::move(CLI::GetParam<arma::mat>("query"));
+  const double bandwidth = CLI::GetParam<double>("bandwidth");
+  const std::string kernelStr = CLI::GetParam<std::string>("kernel");
+  const std::string treeStr = CLI::GetParam<std::string>("tree");
+  const double relError = CLI::GetParam<double>("rel_error");
+  const double absError = CLI::GetParam<double>("abs_error");
+  const bool breadthFirst = CLI::GetParam<bool>("breadth_first");
+  // Initialize results vector.
+  arma::vec estimations;
+
+  // You can only specify reference data or a pre-trained model.
+  RequireOnlyOnePassed({ "reference", "input_model" }, true);
+  ReportIgnoredParam({{ "input_model", true }}, "tree");
+  ReportIgnoredParam({{ "input_model", true }}, "kernel");
+  ReportIgnoredParam({{ "input_model", true }}, "rel_error");
+  ReportIgnoredParam({{ "input_model", true }}, "abs_error");
+  ReportIgnoredParam({{ "input_model", true }}, "breadth_first");
+
+  KDEModel* kde;
+
+  if (CLI::HasParam("reference"))
+  {
+    arma::mat reference = std::move(CLI::GetParam<arma::mat>("reference"));
+
+    kde = new KDEModel();
+    // Set parameters
+    kde->Bandwidth() = bandwidth;
+    kde->RelativeError() = relError;
+    kde->AbsoluteError() = absError;
+    kde->BreadthFirst() = breadthFirst;
+
+    // Set KernelType
+    if (kernelStr == "gaussian")
+      kde->KernelType() = KDEModel::GAUSSIAN_KERNEL;
+    else if (kernelStr == "epanechnikov")
+      kde->KernelType() = KDEModel::EPANECHNIKOV_KERNEL;
+    else
+      Log::Fatal << "Input kernel is not valid or not supported yet."
+                 << std::endl;
```

Comment:
This can cause a subtle memory leak---`Log::Fatal` will terminate, but we should still delete the `kde` object before calling that.  You could also use:

```
RequireParamInSet<string>("kernel", { "gaussian", "epanechnikov" });
```

Reply:
Solved in e97d1bf8c313736948425fa56b3d8de24cc4b4ec